test/workers/delete_plain_object_worker_test.rb

```diff
@@ -54,4 +54,55 @@ def test_perform_destroy_without_association
       refute object.destroyed?
     end
   end
+
+  class StaleObjectErrorTest < DeletePlainObjectWorkerTest
+    module LoadTargetWithFiber
+      # Overriding the `delete` method of HasOneAssociation https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/associations/has_one_association.rb#L53-L64
+      # When entering this method, override the `load_target` method so we can hand over the execution to the main thread
+      def delete
+        def load_target
+          super.tap { Fiber.yield }
+        end
+
+        super
+      end
+    end
+
+    def test_race_condition
+      service = FactoryGirl.create(:simple_service)
+      # There is a restriction on deleting service, at least one should remain
+      FactoryGirl.create(:simple_service, account: service.account)
+
+      proxy = Proxy.find service.proxy.id
```

Comment:
We cannot do `proxy = service.proxy` because it will be caching the `proxy` in the `associaton#target`